### DISCUSSION TITLE
Graphical results

### DISCUSSION THREAD

**Discussion by ottapav** [qpsolvers/qpbenchmark#48] [General]
Hi,
The results generated by this benchmark are pretty extensive already. With minimal additional effort, is it possible to add a plotting feature, e.g., similar to that for Maros Meszaros benchmark in https://github.com/Simple-Robotics/proxqp_benchmark?
![image](https://user-images.githubusercontent.com/10757098/217569245-56cdc73f-f80e-48ec-84d3-8cefa605f49a.png)

**Comment by stephane-caron** [comment]
Good point, which I think was not mentioned before in the benchmark discussions. The plot you are mentioning here is a performance profile. We chose not to plot those in the benchmark for reasons detailed here: https://github.com/stephane-caron/qpsolvers_benchmark/discussions/49

There is however the beginning of a plotting tool with the ``hist`` command, which focuses on a single metric (metrics are runtime, primal residual, duality gap, ...). To take a use case close to a performance profile, we can look at runtime:

![image](https://user-images.githubusercontent.com/1189580/217618342-da2637c6-c2d0-46ae-9071-b0f4e8dee6ca.png)

This plot being generated by this command line:

```
./benchmark.py maros_meszaros_dense hist runtime high_accuracy --solvers proxqp osqp scs --bins 42
```

**Reply by ottapav** [reply]
Good idea! Possible improvement: To show the total number of evaluated problems as the upper bound on the y-axis. It cannot be seen if e.g. proxqp solved all the problems eventually from this figure.

**Reply by ottapav** [reply]
I saw #39, but it is a little bit confusing that in the figure above proxqp solved one problem more than the others.
![image](https://user-images.githubusercontent.com/10757098/217757067-409b06c8-493c-4bb2-9cfd-3ee07f3bbf61.png)

**Reply by stephane-caron** [reply]
You're right, this was actually an issue with the underlying histogram calculation. I implemented a simpler solution in https://github.com/stephane-caron/qpsolvers_benchmark/pull/50 that fixes this.

> Good idea! Possible improvement: To show the total number of evaluated problems as the upper bound on the y-axis. It cannot be seen if e.g. proxqp solved all the problems eventually from this figure.

Done :+1: You can pull the latest commit from ``main`` to get the fix.

Note that I've changed the command name from ``hist`` to just ``plot``:

```
python benchmark.py maros_meszaros_dense plot runtime high_accuracy --solvers proxqp osqp scs
```

The result is now:

![image](https://user-images.githubusercontent.com/1189580/217885413-c3e68c24-d04c-4a76-9ba5-f859899f92e8.png)

Thanks for your feedback!

**Reply by ottapav** [reply]
@stephane-caron Hi Stephan, when you make some changes in the plotting feature, please consider adding a unit for the x-axis (e.g., [s]). And could you make lines distinguishable in the grayscale (maybe by using some markers)? I'm finalizing a paper ;)

**Reply by stephane-caron** [reply]
For adjustments your best bet for now is to edit `plot_metric.py` directly. In general it would be better in benchmark code to separate data aggregation from plotting, which would make it easier to customize plots to taste. Feel free to open a feature request about this.

In the meantime, note that we can also customize the plots after the fact when passing the ``savefig`` argument:

```console
$ ipython -i ./benchmark.py maros_meszaros_dense plot runtime default -- --savefig runtime.pdf
Python 3.8.10 (default, Mar 13 2023, 10:26:41) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.22.0 -- An enhanced Interactive Python. Type '?' for help.
[2023-05-02 10:26:29,479] [info] Loading existing results from ./results/maros_meszaros_dense.csv (results.py:80)

In [1]: from matplotlib import pyplot as plt

In [2]: plt.xlabel("Runtime [s]")
Out[2]: Text(0.5, 18.922222222222217, 'Runtime [s]')

In [3]: plt.show()
```

**Comment by ottapav** [comment]
Hi Stephane,
Just for your information: Currently, I have finished the Python interface
and started (local) integration to the qpsolvers to be able to run the
benchmark with our solver.

Pavel Otta


čt 9. 2. 2023 v 18:03 odesílatel Stéphane Caron ***@***.***>
napsal:

**Reply by stephane-caron** [reply]
Sounds good! Feel free to open a PR when it is ready.
